lib/galaxy/model/store/__init__.py

```diff
@@ -2215,6 +2217,12 @@ def add_dataset_collection(
     def add_dataset(self, dataset: model.DatasetInstance, include_files: bool = True) -> None:
         self.included_datasets[dataset] = (dataset, include_files)
 
+    def _ensure_dataset_file_exists(self, dataset: model.DatasetInstance) -> None:
+        state = dataset.dataset.state
+        if state in [model.Dataset.states.OK] and not dataset.file_name:
```

Comment:
It could be worth logging the object store id as well (dataset.dataset.object_store_id)